### PR TITLE
Polish QuarkusClassLoader

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java [hunk 1]
```diff
@@ -239,19 +240,19 @@ public void addResourceMapping(ClassPathElement classPathElement, String resourc
...
+        public void addBannedResource(String resource) {
             bannedResources.add(resource);
         }
 
         public ClassPathResourceIndex build() {
             Map<String, ClassPathElement[]> compactedResourceMapping = new HashMap<>(resourceMapping.size());
             for (Entry<String, List<ClassPathElement>> resourceMappingEntry : resourceMapping.entrySet()) {
                 compactedResourceMapping.put(resourceMappingEntry.getKey(),
-                        resourceMappingEntry.getValue().toArray(new ClassPathElement[resourceMappingEntry.getValue().size()]));
+                        resourceMappingEntry.getValue().toArray(EMPTY_CLASSPATH_ELEMENT));
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:255]: Do you have a pointer for this? I remembered I was told your thing was better now but I had a look at the code of the JDK the other day and it wasn't that obvious.

Not asking to justify but if you have a pointer handy, I'll be interested in refreshing my memory.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 2]
```diff
@@ -310,12 +310,14 @@ private ClassPathResourceIndex getClassPathResourceIndex() {
 
                     for (ClassPathElement bannedElement : bannedElements) {
                         classPathResourceIndexBuilder.scanClassPathElement(bannedElement,
-                                classPathResourceIndexBuilder::addBannedResource);
+                                (classPathElement, resource) -> classPathResourceIndexBuilder.addBannedResource(
+                                        resource));
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:314]: Yeah so the reason why I had two parameters was to have a consistent interface for anything called with `scanClassPathElement`.

Do you think dropping the parameter is worth adding a real lambda?

**Review by gsmet — APPROVED** [review]
I'm not entirely convinced but feel free to merge :).

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 2 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:314]: My main gripe about this is that the code looked simpler and more consistent the way I did it but I suppose it's in the eyes of the reader.

Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:314]: I personally I have never seen a method parameter being added just so a method reference can be used. I'm not saying  it's bad, it's just very unintuitive.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java (hunk 1 above)
Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:255]: https://shipilev.net/blog/2016/arrays-wisdom-ancients/ is why I use the idiom.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 2 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:314]: Well, if you think in terms of contract of the `scanClassPathElement`, I think it makes kinda sense.

Anyway, probably not worth discussing it.

**Pull request merged**